streams/src/main/java/org/apache/kafka/streams/kstream/Windows.java

```diff
@@ -39,17 +43,48 @@
     private long maintainDurationMs = 24 * 60 * 60 * 1000L; // default: one day
     @Deprecated public int segments = 3;
 
+    private Duration grace;
+
     protected Windows() {}
 
+    /**
+     * Reject late events that arrive more than {@code afterWindowEnd}
+     * after the end of its window.
+     *
+     * @param afterWindowEnd The grace period to admit late-arriving events to a window.
+     * @return this updated builder
+     */
+    public Windows<W> grace(final Duration afterWindowEnd) {
```

Comment:
Personally, I think that `Duration` is more ergonomic and less error-prone.

Also, we don't require the duration to be "in milliseconds", just expressible in milliseconds. I.e., `Duration.ofDays(2)` is fine, but `Duration.ofDays(Long.MAX_VALUE)` is not.

It's heavy-weight to migrate our whole API to use Duration, but I was thinking that we could at least migrate whatever parts we're touching. But I didn't really consider the cost of having a heterogeneous API. I can switch it to `long gracePeriodMs` instead, if you don't favor the gradual transition approach, and maybe we can consider moving the whole API to Duration in a separate discussion.

WDYT @guozhangwang ?

Reply:
I agree with you that `Duration` is a better choice over using a `long` value, as it's much easier to grok what is going on with the parameter.

However, I think that having a consistent API across the board outweighs that issue, especially considering the point raised above by @guozhangwang regarding the Kafka timestamp field and the timestamp-extractor.

I would not have a problem with upgrading the entire API to use `Duration` in one shot,  but I think doing the changes incrementally as we update the code may not be fast enough.  

I say this because without a dedicated plan to make the changes from `long` to `Duration` IMHO there will parts that we miss either from oversight or just because those sections of the code don't need to change.

It's a tough question, where I don't feel there's a wrong answer just what you feel more comfortable with doing.

Reply:
Ok! I'll change it to long.